1417: Implement authentication for user import endpoint

seluianova opened the pull request:
### Short description
Implement authentication for user import endpoint using bearer tokens introduced in https://github.com/digitalfabrik/entitlementcard/pull/1607

### Proposed changes
- add api token validation to UserImportHandler
- adjust tests

### Side effects
no?

### Testing
1. Open administration --> login as a koblenz project admin --> Projekt verwalten --> create token
2. Try to import users using curl or Postman with and without auth header
`curl -H "Authorization: Bearer <my_token>" -F file=@import_testdata_local.csv http://0.0.0.0:8000/users/import`

**CSV content:**
```
regionKey,userHash,startDate,endDate,revoked
07111,"$argon2id$v=19$m=19456,t=2,p=1$cr3lP9IMUKNz4BLfPGlAOHq1z98G5/2tTbhDIko35tY",01.01.2024,01.01.2025,false
07111,"$argon2id$v=19$m=19456,t=2,p=1$POG+rDhWlxSW7ItkP9hZRHqr46u6a6TQO1Ib2X+khMA",01.01.2024,01.01.2025,false
07111,"$argon2id$v=19$m=19456,t=2,p=1$zf7vqckgmYBPbjjW1CwqudF+OjHmo2EifgxhuuVHFs0",01.01.2024,01.01.2025,true
```

### Resolved issues
Fixes: #1417

seluianova commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/userdata/webservice/UserImportHandler.kt:
```diff
@@ -37,15 +49,33 @@ class UserImportHandler {
+        return transaction {
+            ApiTokenEntity.find { ApiTokens.expirationDate greater LocalDate.now() }
+                .firstOrNull { BCrypt.verifyer().verify(token.toCharArray(), it.token).verified }
+                ?: throw ForbiddenException()
```
It turned out that the current hashing does not allow selecting a token by hash from the database, because the hashing function returns a different result for the same token value every time.
To validate a token we need to get 2 values:
- token from auth header
- hash value from the database
So to find the right token in the database, I select all the values and perform the check on each one.

f1sh1918 reviewed (commented):
Works as expected. nicely done! 👍 
Tested with curl and postman

I just had to change my curl command, because `-header` didn't work
```js
curl -H "Authorization: Bearer <my_token>" -F file=@import_testdata_local.csv http://0.0.0.0:8000/users/import
```

One thing that we might change here to also be able to set staging environment to `production: true` again is that we may get rid of this.
```js
   // TODO remove the condition once the Koblenz project is ready to go live
        if (!production) {
            app.post("/users/import") { ctx -> userImportHandler.handle(ctx) }
        }
```

Since the endpoint is now protected, we don't need this anymore

seluianova commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/userdata/webservice/UserImportHandler.kt:
```diff
@@ -19,15 +26,22 @@ import java.time.LocalDate
+            val projectConfig = backendConfiguration.getProjectConfig(project.project)
+
+            if (!projectConfig.selfServiceEnabled) {
+                throw UserImportException("User import is not enabled in the project")
```
I feel like '400 bad request' is more appropriate actually 🤔
5xx are server-side errors, but when we disabled a feature in the config, it is not a server-side error.

f1sh1918 replied:
hm okay but the "Server nicht erreichbar" error message is even not very specific and leads to confusion on our side

seluianova replied:
it shouldn't be "Server nicht erreichbar " currently. do you get this error?

f1sh1918 commented:
Please set `selfServiceEnabled: true` fpr koblenz in the `config.yml`

michael-markl commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/database/repos/ApiTokensRepository.kt:
```diff
@@ -18,4 +19,8 @@ object ApiTokensRepository {
     }
+
+    fun findByToken(token: ByteArray): ApiTokenEntity? {
+        return ApiTokenEntity.find { (ApiTokens.token eq token) }.singleOrNull()
```
Probably, we should add an index for the token column of ApiTokens then?
And maybe we should rename the column to `tokenHash`?

michael-markl commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/auth/webservice/schema/ApiTokenService.kt:
```diff
@@ -32,13 +32,11 @@ class ApiTokenService {
+        val token = Base64.getEncoder().encodeToString(bytes)
 
-        val encryptedToken = BCrypt.withDefaults().hash(11, token.toCharArray())
+        val encryptedToken = PasswordCrypto.hashWithSHA256(token.toByteArray())
```
It's not encrypted, it's just hashed, right?
```suggestion
        val tokenHash = PasswordCrypto.hashWithSHA256(token.toByteArray())
```

seluianova commented:
**UPD**:
* set selfServiceEnabled: true for koblenz in the config.yml
* renamed token column to tokenHash
* added unique index for tokenHash column

f1sh1918 commented:
> 1. https://git.tuerantuer.org/DF/salt/pulls

i don't think we need this on production now. I will give koblenz credentials for staging. We may create an issue for that to not forget to unable it in the end

f1sh1918 reviewed (approved):
Tested again. Works fine

michael-markl commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/migration/migrations/V0020_AddApiTokenTable.kt:
```diff
@@ -9,13 +9,15 @@ internal class V0020_AddApiTokenTable : Migration() {
                 ALTER TABLE apitokens ADD CONSTRAINT fk_apitokens_creator__id FOREIGN KEY ("creatorId") REFERENCES administrators(id) ON DELETE RESTRICT ON UPDATE RESTRICT;
                 ALTER TABLE apitokens ADD CONSTRAINT fk_apitokens_projectid__id FOREIGN KEY ("projectId") REFERENCES projects(id) ON DELETE RESTRICT ON UPDATE RESTRICT;
+                
+                ALTER TABLE ONLY apitokens ADD CONSTRAINT apitokens_tokenhash_unique UNIQUE ("tokenHash");
```
It seems this migration was already released in version [2024.9.0](https://github.com/digitalfabrik/entitlementcard/blob/2024.9.0-all/backend/src/main/kotlin/app/ehrenamtskarte/backend/migration/migrations/V0020_AddApiTokenTable.kt) ? If so, we should rather make these changes in a new migration? (Maybe I'm missing context though)

f1sh1918 replied:
https://github.com/digitalfabrik/entitlementcard/issues/1645

f1sh1918 replied:
just to ensure here a the current table list
```js
 List of relations
     Schema      |              Name               |   Type   |  Owner
-----------------+---------------------------------+----------+----------
 entitlementcard | acceptingstores                 | table    | backend
 entitlementcard | acceptingstores_id_seq          | sequence | backend
 entitlementcard | addresses                       | table    | backend
 entitlementcard | addresses_id_seq                | sequence | backend
 entitlementcard | administrators                  | table    | backend
 entitlementcard | administrators_id_seq           | sequence | backend
 entitlementcard | applications                    | table    | backend
 entitlementcard | applications_id_seq             | sequence | backend
 entitlementcard | applicationverifications        | table    | backend
 entitlementcard | applicationverifications_id_seq | sequence | backend
 entitlementcard | cards                           | table    | backend
 entitlementcard | cards_id_seq                    | sequence | backend
 entitlementcard | categories                      | table    | backend
 entitlementcard | categories_id_seq               | sequence | backend
 entitlementcard | contacts                        | table    | backend
 entitlementcard | contacts_id_seq                 | sequence | backend
 entitlementcard | geography_columns               | view     | postgres
 entitlementcard | geometry_columns                | view     | postgres
 entitlementcard | migrations                      | table    | backend
 entitlementcard | physicalstores                  | table    | backend
 entitlementcard | physicalstores_id_seq           | sequence | backend
 entitlementcard | projects                        | table    | backend
 entitlementcard | projects_id_seq                 | sequence | backend
 entitlementcard | regions                         | table    | backend
 entitlementcard | regions_id_seq                  | sequence | backend
 entitlementcard | spatial_ref_sys                 | table    | postgres
 entitlementcard | userentitlements                | table    | backend
 entitlementcard | userentitlements_id_seq         | sequence | backend
```

michael-markl replied:
I guess for the future, it should usually never be a problem to just add another migration and should usually be preferred over adjusting previous ones, because that is dangerous territory.

f1sh1918 replied:
yes i mean we can can even start here

seluianova replied:
moved to the separate migration

steffenkleinle commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/common/webservice/WebService.kt:
```diff
@@ -77,10 +77,7 @@ class WebService {
-        if (!production) {
-            app.post("/users/import") { ctx -> userImportHandler.handle(ctx) }
-        }
+        app.post("/users/import") { ctx -> userImportHandler.handle(ctx) }
```
:thought_balloon: Personally I dislike abbreviations like this, as it makes it harder to read if you are not familiar with specific terms or the project in general. Especially in kotlin, since there is the default `it` available which makes the code more concise.
```suggestion
        app.post("/users/import") { userImportHandler.handle(it) }
```
If explicit naming adds anything or is necessary due to shadowing, the full word should be used imo:
```suggestion
        app.post("/users/import") { context -> userImportHandler.handle(context) }
```
Since this is nothing you introduced in this PR, feel free to ignore, just wanted to dump this somewhere :D

seluianova replied:
I'm too lazy to change it here, but I'll update it in my other pull-request, along with other comments :)

seluianova replied:
I'm too lazy to change it here, but I'll push that in my other pull-request, together with some other requested changes

steffenkleinle commented on backend/src/main/kotlin/app/ehrenamtskarte/backend/userdata/webservice/UserImportHandler.kt:
```diff
@@ -37,15 +51,32 @@ class UserImportHandler {
+        }
+    }
+
+    private fun authenticate(ctx: Context): ApiTokenEntity {
```
:upside_down_face: See comment before.
```suggestion
    private fun authenticate(context: Context): ApiTokenEntity {
```

steffenkleinle reviewed (approved):
Tested on firefox, LGTM.
